store/src/java/com/zimbra/cs/service/admin/CheckBlobConsistency.java

```diff
@@ -65,63 +67,82 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
+        Map<StoreType, StoreManager> storeManagerMap = new HashMap<>();
+        for (Volume volume: vols) {
+            StoreManager sm = StoreManager.getReaderSMInstance(volume.getId());
+            if (sm instanceof FileBlobStore && volumeElementList.isEmpty()) {
+                storeManagerMap.putIfAbsent(StoreType.INTERNAL, sm);
+                switch (volume.getType()) {
+                    case Volume.TYPE_MESSAGE:
+                    case Volume.TYPE_MESSAGE_SECONDARY:
+                        volumeIds.add(volume.getId());
+                        break;
                 }
+            } else if (sm instanceof ExternalStoreManager) {
+                storeManagerMap.putIfAbsent(StoreType.EXTERNAL, sm);
             } else {
-                // Read volume id's from the request.
-                for (Element volumeEl : volumeElementList) {
-                    short volumeId = (short) volumeEl.getAttributeLong(AdminConstants.A_ID);
-                    Volume vol = VolumeManager.getInstance().getVolume(volumeId);
-                    if (vol.getType() == Volume.TYPE_INDEX) {
-                        throw ServiceException.INVALID_REQUEST("Index volume " + volumeId + " is not supported", null);
-                    } else {
-                        volumeIds.add(volumeId);
+                storeManagerMap.putIfAbsent(StoreType.NOT_SUPPORTED, sm);
+            }
+        }
+        for (StoreManager sm : storeManagerMap.values()) {
+            if (sm instanceof ExternalStoreManager && volumeElementList.isEmpty()) {
+                for (int mboxId : mailboxIds) {
+                    ExternalBlobConsistencyChecker checker = new ExternalBlobConsistencyChecker();
+                    checker.setStoreManager(sm);
+                    BlobConsistencyChecker.Results results = checker.check(null, mboxId, checkSize, reportUsedBlobs);
+                    if (results.hasInconsistency() || reportUsedBlobs) { //or checking used blobs
+                        Element mboxEl = response.addElement(AdminConstants.E_MAILBOX).addAttribute(AdminConstants.A_ID, mboxId);
+                        results.toElement(mboxEl);
                     }
                 }
-            }
-
-            for (int mboxId : mailboxIds) {
-                BlobConsistencyChecker checker = new BlobConsistencyChecker();
-                BlobConsistencyChecker.Results results = checker.check(volumeIds, mboxId, checkSize, reportUsedBlobs);
-                if (results.hasInconsistency() || reportUsedBlobs) {
-                    Element mboxEl = response.addElement(AdminConstants.E_MAILBOX).addAttribute(AdminConstants.A_ID, mboxId);
-                    results.toElement(mboxEl);
+            } else if (sm instanceof FileBlobStore) {
+                for (int mboxId : mailboxIds) {
+                    BlobConsistencyChecker checker = new BlobConsistencyChecker();
+                    checker.setStoreManager(sm);
```

Comment:
The lines from 116 to 120 is same as 106 to 110. We can create one method for these lines of codes and use instanceOf for checker class.